modules/core/src/test/java/org/apache/ignite/internal/util/IgniteUtilsSelfTest.java

```diff
@@ -1296,6 +1303,95 @@ public void testDoInParallelException() {
         }
     }
 
+    /**
+     *
+     */
+    @Test
+    public void testGridConcurrentMultiPairQueueCorrectness() throws Exception {
+        Integer[] arr2 = new Integer[]{2, 4};
+        Integer[] arr1 = new Integer[]{1, 3, 5, 7, 9, 11, 13, 15, 17, 19};
+        Integer[] arr4 = new Integer[]{};
+        Integer[] arr5 = new Integer[]{};
+        Integer[] arr3 = new Integer[]{100, 200, 300, 400, 500, 600, 600, 700};
+        Integer[] arr6 = new Integer[]{};
+
+        Collection<T2<Integer, Integer[]>> keyWithArr = new HashSet<>();
+
+        final Map<Integer, Collection<Integer>> mapForCheck = new ConcurrentHashMap<>();
+
+        final Map<Integer, Collection<Integer>> mapForCheck2 = new ConcurrentHashMap<>();
+
+        keyWithArr.add(new T2<>(10, arr2));
+        keyWithArr.add(new T2<>(20, arr1));
+        keyWithArr.add(new T2<>(30, arr4));
+        keyWithArr.add(new T2<>(40, arr5));
+        keyWithArr.add(new T2<>(50, arr3));
+        keyWithArr.add(new T2<>(60, arr6));
+
+        mapForCheck.put(10, Collections.synchronizedCollection(new ArrayList<>(Arrays.asList(arr2))));
+        mapForCheck.put(20, Collections.synchronizedCollection(new ArrayList<>(Arrays.asList(arr1))));
+        mapForCheck.put(50, Collections.synchronizedCollection(new ArrayList<>(Arrays.asList(arr3))));
+
+        mapForCheck2.put(10, Collections.synchronizedCollection(new ArrayList<>(Arrays.asList(arr2))));
+        mapForCheck2.put(20, Collections.synchronizedCollection(new ArrayList<>(Arrays.asList(arr1))));
+        mapForCheck2.put(50, Collections.synchronizedCollection(new ArrayList<>(Arrays.asList(arr3))));
+
+        final GridConcurrentMultiPairQueue<Integer, Integer> queue = new GridConcurrentMultiPairQueue<>(keyWithArr);
+
+        GridTestUtils.runMultiThreaded(() -> {
+            T2<Integer, Integer> pair = queue.poll();
+
+            while (pair != null) {
+                assertTrue(mapForCheck.containsKey(pair.get1()));
+
+                mapForCheck.get(pair.get1()).remove(pair.get2());
+
+                if (mapForCheck.get(pair.get1()).isEmpty())
```

Comment:
You can fail with NPE here, collection might have been removed by another thread. Please be careful.